modules/light-clients/08-wasm/types/proposal_handle_test.go

```diff
@@ -78,6 +83,109 @@ func (suite *TypesTestSuite) TestCheckSubstituteAndUpdateStateGrandpa() {
 	}
 }
 
+func (suite *TypesTestSuite) TestCheckSubstituteAndUpdateState() {
+	var substituteClientState exported.ClientState
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expErr   error
+	}{
+		{
+			"success",
+			func() {
+				suite.mockVM.RegisterSudoCallback(
+					types.MigrateClientStoreMsg{},
+					func(_ cosmwasm.Checksum, _ wasmvmtypes.Env, sudoMsg []byte, store cosmwasm.KVStore, _ cosmwasm.GoAPI, _ cosmwasm.Querier, _ cosmwasm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+						var payload types.SudoMsg
+						err := json.Unmarshal(sudoMsg, &payload)
+						suite.Require().NoError(err)
+
+						suite.Require().NotNil(payload.MigrateClientStore)
+						suite.Require().Nil(payload.UpdateState)
+						suite.Require().Nil(payload.UpdateStateOnMisbehaviour)
+						suite.Require().Nil(payload.VerifyMembership)
+						suite.Require().Nil(payload.VerifyNonMembership)
+						suite.Require().Nil(payload.VerifyUpgradeAndUpdateState)
+
+						bz, err := json.Marshal(types.EmptyResult{})
+						suite.Require().NoError(err)
+
+						return &wasmvmtypes.Response{Data: bz}, wasmtesting.DefaultGasUsed, nil
+					},
+				)
+			},
+			nil,
+		},
+		{
+			"failure: invalid substitute client state",
+			func() {
+				substituteClientState = &ibctm.ClientState{}
+			},
+			clienttypes.ErrInvalidClient,
+		},
+		{
+			"failure: code hashes do not match",
+			func() {
+				substituteClientState = &types.ClientState{
+					CodeHash: []byte("invalid"),
+				}
+			},
+			clienttypes.ErrInvalidClient,
+		},
+		{
+			"failure: contract returns error",
+			func() {
+				suite.mockVM.RegisterSudoCallback(
+					types.MigrateClientStoreMsg{},
+					func(_ cosmwasm.Checksum, _ wasmvmtypes.Env, _ []byte, _ cosmwasm.KVStore, _ cosmwasm.GoAPI, _ cosmwasm.Querier, _ cosmwasm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+						return nil, wasmtesting.DefaultGasUsed, wasmtesting.ErrMockContract
+					},
+				)
+			},
+			types.ErrWasmContractCallFailed,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupWasmWithMockVM()
+
+			endpointA := wasmtesting.NewWasmEndpoint(suite.chainA)
+			err := endpointA.CreateClient()
+			suite.Require().NoError(err)
+
+			subjectClientStore := suite.chainA.App.GetIBCKeeper().ClientKeeper.ClientStore(suite.chainA.GetContext(), endpointA.ClientID)
+			subjectClientState := endpointA.GetClientState()
+
+			endpointB := wasmtesting.NewWasmEndpoint(suite.chainA)
```

Comment:
maybe substituteEndpoint? thought endpointB was supposed to refer to an endpoint on chainB

Reply:
ah yea, solid point